## Data Analysis Process

1. Asking Questions
2. Data Wrangling<br>
    a. Gathering Data 
        - i. CSV files 
        - ii. APIs 
        - iii. Web Scraping 
        - iv. Databases 
    b. Assessing Data<br>
    c. Cleaning Data
3. Exploratory Data Analysis
4. Drawing Conclusion
5. Comunicating Results

### 1. Data Wrangling

- Data Gathering
- Data Accessing
- Data Cleaning

1b. Data Accessing

In this step, the data is to be understood more deeply. Before implementing methods to clean it, you will definitely need to have a better idea about what the data is about.

Types of Unclean Data

There are 2 kinds of unclean data

### Dirty Data (Data with Quality issues): Dirty data, also known as low quality data. Low quality data has content issues.

- Duplicated data
- Missing Data
- Corrupt Data
- Inaccurate Data
### Messy Data (Data with tidiness issues): Messy data, also known as untidy data. Untidy data has structural issues.Tidy data has the following properties:

- Each variable forms a column
- Each observation forms a row
- Each observational unit forms a table

In [2]:
import pandas as pd
import numpy as np

In [3]:
patients=pd.read_csv('patients.csv')
treatment=pd.read_csv('treatments.csv')
adverse=pd.read_csv('adverse_reactions.csv')
treatments_cut = pd.read_csv('treatments_cut.csv')

In [4]:
print(patients.shape)
print(treatment.shape)
print(adverse.shape)

(503, 14)
(280, 7)
(34, 3)


In [5]:
patients.head(5)

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [6]:
treatment.head(5)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [14]:
adverse.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [9]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [12]:
treatment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [13]:
adverse.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [10]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [14]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [15]:
treatment.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [19]:
patients.sample(5)

patient_id assigned_sex given_name     surname             address  \
296         297       female        Chỉ         Lâm                 NaN   
13           14       female   Anenechi       Chidi    826 Broad Street   
177         178       female     Monika      Lončar   377 Norman Street   
5             6         male     Rafael       Costa  1140 Willis Avenue   
457         458       female      Mevla  Grabljevec   2645 Moore Avenue   

              city    state  zip_code        country  \
296            NaN      NaN       NaN            NaN   
13      Birmingham       AL   35203.0  United States   
177    Los Angeles       CA   90014.0  United States   
5    Daytona Beach  Florida   32114.0  United States   
457     Fort Worth       TX   76102.0  United States   

                                        contact  birthdate  weight  height  \
296                                         NaN  5/14/1990   181.1      63   
13   AnenechiChidi@armyspy.com+1 (205) 417-8095   3/7/1961   228.4      67   
177          323-327-4721MonikaLoncar@rhyta.com  7/19/1927   203.7      63   
5      386-334-5237RafaelCardosoCosta@gustr.com  8/31/1931   183.9      70   
457     817-777-5289MevlaGrabljevec@teleworm.us  12/2/1984   215.8      66   

      bmi  
296  32.1  
13   35.8  
177  36.1  
5    26.4  
457  34.8

### 1. Write a summary for your data

This is a dataset about 500 patients of which 350 patients participated in a clinical trial. None of the patients were using Novodra (a popular injectable insulin) or Auralin (the oral insulin being researched) as their primary source of insulin before. All were experiencing elevated HbA1c levels.

All 350 patients were treated with Novodra to establish a baseline HbA1c level and insulin dose. After 4 weeks, which isn’t enough time to capture all the change in HbA1c that can be attributed by the switch to Auralin or Novodra:

- 175 patients switched to Auralin for 24 weeks
- 175 patients continued using Novodra for 24 weeks

Data about patients feeling some adverse effects is also recorded.

### 2. Write Column descriptions

#### **Table** -> `patients`:

- `patient_id`: the unique identifier for each patient in the Master Patient Index (i.e. patient database) of the pharmaceutical company that is producing Auralin
- `assigned_sex`: the assigned sex of each patient at birth (male or female)
- `given_name`: the given name (i.e. first name) of each patient
- `surname`: the surname (i.e. last name) of each patient
- `address`: the main address for each patient
- `city`: the corresponding city for the main address of each patient
- `state`: the corresponding state for the main address of each patient
- `zip_code`: the corresponding zip code for the main address of each patient
- `country`: the corresponding country for the main address of each patient (all United states for this clinical trial)
- `contact`: phone number and email information for each patient
- `birthdate`: the date of birth of each patient (month/day/year). The inclusion criteria for this clinical trial is age >= 18 (there is no maximum age because diabetes is a growing problem among the elderly population)
- `weight`: the weight of each patient in pounds (lbs)
- `height`: the height of each patient in inches (in)
- `bmi`: the Body Mass Index (BMI) of each patient. BMI is a simple calculation using a person's height and weight. The formula is BMI = kg/m2 where kg is a person's weight in kilograms and m2 is their height in metres squared. A BMI of 25.0 or more is overweight, while the healthy range is 18.5 to 24.9. The inclusion criteria for this clinical trial is 16 >= BMI >= 38.

#### **Table** -> `treatments` and `treatment_cut`:

- `given_name`: the given name of each patient in the Master Patient Index that took part in the clinical trial
- `surname`: the surname of each patient in the Master Patient Index that took part in the clinical trial
- `auralin`: the baseline median daily dose of insulin from the week prior to switching to Auralin (the number before the dash) and the ending median daily dose of insulin at the end of the 24 weeks of treatment measured over the 24th week of treatment (the number after the dash). Both are measured in units (shortform 'u'), which is the international unit of measurement and the standard measurement for insulin.
- `novodra`: same as above, except for patients that continued treatment with Novodra
- `hba1c_start`: the patient's HbA1c level at the beginning of the first week of treatment. HbA1c stands for Hemoglobin A1c. The HbA1c test measures what the average blood sugar has been over the past three months. It is thus a powerful way to get an overall sense of how well diabetes has been controlled. Everyone with diabetes should have this test 2 to 4 times per year. Measured in %.
- `hba1c_end`: the patient's HbA1c level at the end of the last week of treatment
- `hba1c_change`: the change in the patient's HbA1c level from the start of treatment to the end, i.e., hba1c_start - hba1c_end. For Auralin to be deemed effective, it must be "noninferior" to Novodra, the current standard for insulin. This "noninferiority" is statistically defined as the upper bound of the 95% confidence interval being less than 0.4% for the difference between the mean HbA1c changes for Novodra and Auralin (i.e. Novodra minus Auralin).

#### **Table** -> `adverse_reactions`

- `given_name`: the given name of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra)
- `surname`: the surname of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra)
- `adverse_reaction`: the adverse reaction reported by the patient

### Types of Assessment
There are 2 types of assessment styles

- `Manual` - Looking through the data manually in google sheets
- `Programmatic` - By using pandas functions such as info(), describe() or sample()

### Steps in Assessment
There are 2 steps involved in Assessment

- Discover
- Document

### Issues with the dataset

1. Dirty Data

  Table - `Patients`

    - patient_id = 9 has misspelled name 'Dsvid' instead of David `accuracy`
    - state col sometimes contain full name and some times abbrivietation `consistency`
    - zip code col has entries with 4 digit `validity`
    - data missing for 12 patients in address,city,	state,zip_code	,country,	contact `completion`
    - incorrect data type assigned to sex, zip code, birthdate `validity`
    - duplicate entries by the name of John Doe `accuracy`
    - one patient has weight = 48 pounds `accuracy`
    - one patient has height = 27 inches `accuracy`
Table - `Treatments` & `Treatments_cut`

    - given_name and surname col is is all lower case `consistency`
    - remove u from Auralin and Novadra cols `validity`
    - '-' in novadra and Auralin col treated as nan `validity`
    - missing values in hba1c_change col `completion`
    - 1 duplicate entry by the name Joseph day `accuracy`
    - in hba1c_change 9 instead of 4 `accuracy`

  Table - `Adverse_reactions`

    - given_name and surname are all in lower case `consistency`

2. Messy Data

  Table - `Patients`

    - contact col contains both phone and email

  Table - `Treatments` & `Treatments_cut`
  - Auralin and Novadra col should be split into 2 cols start and end dose
    - merge both the tables

  Table - `Adverse_reactions`

    - This table should not exist independently

### Automatic Assessment

- head and tail
- sample
- info
- isnull
- duplicated
- describe

In [11]:
adverse.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [24]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [14]:
treatment[treatment.duplicated()]

given_name surname    auralin novodra  hba1c_start  hba1c_end  \
136     joseph     day  29u - 36u       -          7.7       7.19   

     hba1c_change  
136           NaN

In [25]:
patients.duplicated().sum()

0

In [26]:
patients['patient_id'].duplicated().sum()

0

In [30]:
patients.duplicated(subset=('given_name','surname')).sum()

5

In [22]:
patients[patients['address'].duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [15]:
treatment[treatment.duplicated(subset=['given_name'	,'surname'])]

given_name surname    auralin novodra  hba1c_start  hba1c_end  \
136     joseph     day  29u - 36u       -          7.7       7.19   

     hba1c_change  
136           NaN

In [16]:
treatments_cut[treatments_cut.duplicated(subset=['given_name'	,'surname'])]

Empty DataFrame
Columns: [given_name, surname, auralin, novodra, hba1c_start, hba1c_end, hba1c_change]
Index: []

In [18]:
adverse.duplicated().sum()

0

In [19]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [20]:
patients[patients['height'] == 27]

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state  zip_code        country                          contact  birthdate  \
4    AL   36303.0  United States  334-515-7487TimNeudorf@cuvox.de  2/18/1928   

   weight  height   bmi  
4   192.3      27  26.1

In [31]:
patients[patients['weight'] == 48.800000]

patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state  zip_code        country  \
210  Wooster    OH   44691.0  United States   

                                       contact   birthdate  weight  height  \
210  330-202-2145CamillaZaitseva@superrito.com  11/26/1938    48.8      63   

      bmi  
210  19.1

In [22]:
treatments_cut.describe()

hba1c_start  hba1c_end  hba1c_change
count    70.000000  70.000000     42.000000
mean      7.838000   7.443143      0.518810
std       0.423007   0.418706      0.270719
min       7.510000   7.020000      0.280000
25%       7.640000   7.232500      0.340000
50%       7.730000   7.345000      0.370000
75%       7.860000   7.467500      0.907500
max       9.910000   9.460000      0.970000

In [23]:
treatment.sort_values('hba1c_change',na_position='first')

given_name      surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika     jindrová  41u - 48u          -         7.63       7.20   
2     yukitaka     takenaka          -  39u - 36u         7.68       7.25   
8        saber       ménard          -  54u - 54u         8.08       7.70   
9         asia      woźniak  30u - 36u          -         7.76       7.37   
10      joseph          day  29u - 36u          -         7.70       7.19   
..         ...          ...        ...        ...          ...        ...   
49     jackson      addison          -  42u - 42u         7.99       7.51   
17        gina         cain          -  36u - 36u         7.88       7.40   
32       laura  ehrlichmann          -  43u - 40u         7.95       7.46   
245         wu         sung          -  47u - 48u         7.61       7.12   
138    giovana        rocha          -  23u - 21u         7.87       7.38   

     hba1c_change  
0             NaN  
2             NaN  
8             NaN  
9             NaN  
10            NaN  
..            ...  
49           0.98  
17           0.98  
32           0.99  
245          0.99  
138          0.99  

[280 rows x 7 columns]

### Quality

[Completeness, Validity, Accuracy, Consistency]

#### `Patients`

- Zip code col is stored in decimal format[V]
- Certain zip codes of 4 digits[V] 
- For the state col sometimes full names are use and other times short forms are used[IC]
- Missing values for contact details of 12 patients[C]
- Incorrect data types for assigned_sex and birthdate cols[?]
- Erroneous data for weights and heights[Accuracy]
- Inconsistent way of storing phone numbers in contact col[IC]
- Duplicate and not useful data in dataset[IC]

#### `Treatments`

- hba1c_change col has missing values[C]
- Novodra and Auralin cols has u attached to numbers[V]
- Only 280 rows are there, instead of 350[C]
- Name and surname col is in lower case[IC]
- Incorrect data type for auralin and novadra cols[V]
- hba1c_change 9 instead of 4[Accuracy]
- use of dashes instead of Nan in aurolin and novodra cols[IC]

#### `Adverse Reactions`

- Name and surname col is in lower case[IC]

### Tidiness

#### `Patients`

- Phone and email together in one col i.e. contact

#### `Treatments`

- 3 variables stored in auralin and novodra cols

#### `Adverse Reactions`

- No point of using this table

### Data Quality Dimensions

- Completeness -> is data missing?
- Validity -> is data invalid -> negative height -> duplicate patient id
- Accuracy -> data is valid but not accurate -> weight -> 1kg
- Consistency -> both valid and accurate but written differently -> New Youk and NY

### Order of severity

Completeness <- Validity <- Accuracy <- Consistency

### Data Cleaning Order

1. Quality -> Completeness
2. Tidiness
3. Quality -> Validity
4. Quality -> Accuracy
5. Quality -> Consistency

#### Steps involved in Data cleaning
- Define
- Code
- Test

`Always make sure to create a copy of your pandas dataframe before you start the cleaning process`


# Cleaning Data

You should follow the order provided below:

- Completeness Issues
- Tidiness Issues
- Remaining Data quality issues like validity, accuracy and consistency

### Steps involved in Data cleaning

- Define
- Code 
- Test

`Always make sure to create a copy of your pandas dataframe before you start the cleaning process`

In [36]:
patients_df = patients.copy()
treatment_df = treatment.copy()
treatments_cut_df = treatments_cut.copy()
adverse_reactions_df = adverse.copy()

### Define

- replace all missing values of patients df with no data
- sub hba1c_start from hba1c_end to get all the change values
- in patients table we will use regex to separate email and phone

In [37]:
# code
patients_df.fillna('No data',inplace=True)

In [38]:
# test
patients_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       503 non-null    object 
 5   city          503 non-null    object 
 6   state         503 non-null    object 
 7   zip_code      503 non-null    object 
 8   country       503 non-null    object 
 9   contact       503 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(2), int64(2), object(10)
memory usage: 55.1+ KB


In [41]:
# code 
treatment_df['hba1c_change'] = treatment_df['hba1c_start'] - treatment_df['hba1c_end']
treatments_cut_df['hba1c_change'] = treatments_cut_df['hba1c_start'] - treatments_cut_df['hba1c_end']

In [47]:
patients_df.drop(columns='contact',inplace=True)

In [42]:
# test
treatments_cut_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70 entries, 0 to 69
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    70 non-null     object 
 1   surname       70 non-null     object 
 2   auralin       70 non-null     object 
 3   novodra       70 non-null     object 
 4   hba1c_start   70 non-null     float64
 5   hba1c_end     70 non-null     float64
 6   hba1c_change  70 non-null     float64
dtypes: float64(3), object(4)
memory usage: 4.0+ KB


In [43]:
def find_contact_details(text: str) -> tuple:
    # it the value is NaN, then return it
    if pd.isna(text):
        return np.nan
    
    # create the phone number pattern
    phone_number_pattern = re.compile(r"(\+[\d]{1,3}\s)?(\(?[\d]{3}\)?\s?-?[\d]{3}\s?-?[\d]{4})")
    # find the phone number from the value/text, as a result we will get a list
    phone_number  = re.findall(phone_number_pattern, text)

    # if length is 0, then the regex can't find any ph number, then define with NaN
    if len(phone_number) <= 0:
        phone_number = np.nan
    # if the country code is attached with the ph number, for that case, the first
    # element will be the country code and the 2nd element will be the actual ph
    # number. So, get that ph number
    elif len(phone_number) >= 2:
        phone_number = phone_number[1]
    # else, we will get the ph number. Grab it.
    else:
        phone_number = phone_number[0]

    # if we found the ph number (with/without country code), then remove that part from the actual value.
    # after removing the ph number, the remaining string might be the email address.
    possible_email_add = re.sub(phone_number_pattern, "", text).strip()

    # then return the ph number and the email address
    return phone_number, possible_email_add

In [45]:
import re

In [46]:
patients_df['phone'] = patients_df["contact"].apply(lambda x: find_contact_details(x)).apply(lambda x:x[0])
patients_df['email'] = patients_df["contact"].apply(lambda x: find_contact_details(x)).apply(lambda x:x[1])

In [9]:
patients_copy[patient_copy['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

## a. Completeness Issues

- Missing values for contact details of 12 patients
- Only 280 rows are there, instead of 350
- hba1c_change col has missing values

#### Define

In [ ]:
# Subtract hba1c_start from hba1c_end to get hba1c_change

#### Code

In [10]:
treatment_copy['hba1c_change']=treatment_copy['hba1c_start']-treatment_copy['hba1c_end']

#### Test

In [16]:
treatment_copy.info()

treatment_copy.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    280 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    280.000000
mean      7.985929    7.589286      0.396643
std       0.568638    0.569672      0.060168
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.350000
50%       7.800000    7.420000      0.390000
75%       7.970000    7.570000      0.440000
max       9.950000    9.580000      0.540000

## b. Tidiness Issues

- Phone + email in contact col `patients`
- Novodra and Aurolin col `treatment`
- Merge the `adverse` table to `treatment` table

#### Define

In [36]:
# Novodra and Aurolin using the melt function

#### Code

In [38]:
treatment_copy=pd.melt(treatment_copy, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')

In [40]:
treatment_copy=treatment_copy[treatment_copy['dose']!='-']

In [44]:
treatment_copy['start'],treatment_copy['end']=treatment_copy['dose'].str.split('-').str

In [47]:
treatment_copy.drop(columns={'dose'}, inplace=True)

#### Test

In [48]:
treatment_copy.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

  start   end  
0  41u    48u  
3  33u    36u  
6  37u    42u  
7  31u    38u  
9  30u    36u

### What is Assessing Data?

The data analysis process consists of the following steps

1. Asking Questions
2. Data Wrangling<br><br>
    a. Gathering Data
            i.   CSV files
            ii.  APIs
            iii. Web Scraping
            iv.  Databases
    b. <b>Assessing Data</b><br>
    c. Cleaning Data<br><br>
3. Exploratory Data Analysis
4. Drawing Conclusion
5. Comunicating Results
    

### Assessing Data

In this step, the data is to be understood more deeply. Before implementing methods to clean it, you will definitely need to have a better idea about what the data is about.

### Types of Unclean Data

There are 2 kinds of unclean data

1. <u>Dirty Data</u> (Data with Quality issues):<br>
    Dirty data, also known as low quality data. Low quality data has content issues.<br>
2. <u>Messy Data</u> (Data with tidiness issues)<br>
    Messy data, also known as untidy data. Untidy data has structural issues.

### 1. Dirty Data

When the data has following kinds of issues:

1. Duplicated data
2. Missing Data
3. Corrupt Data
4. Inaccurate Data

This kind of data is known as Dirty data

### 2. Messy Data

When the data has structural issue is it known as Messy or untidy data.

Tidy data has the following properties:

1. Each variable forms a column
2. Each observation forms a row
3. Each observational unit forms a table

If a data violates any one of the above it is known as untidy data

Example:

<img src="untidy.png">

### Types of Assessment

There are 2 types of assessment styles

1. Manual - Looking through the data manually in google sheets
2. Programmatic - By using pandas functions such as info(), describe() or sample()

### Steps in Assessment

There are 2 steps involved in Assessment

1. Discover
2. Document

### Data Quality Dimensions

1. Completeness Issues
2. Validity Issues
3. Accuracy Issues
4. Consistency Issues